### DISCUSSION TITLE
I'm having problems with «$dnstype=~CNAME»

### DISCUSSION THREAD

**Discussion by DandelionSprout** [AdguardTeam/AdGuardHome#4230] [Q&A]
As good as I usually am with adblocker tools' syntaxes, I've encountered a fairly big problem where `$dnstype=~CNAME` isn't being recognised properly: 
![image](https://user-images.githubusercontent.com/22780683/151703766-5a3a4fd7-d274-4c6a-a238-c7352026931a.png)

There's a lot of double negatives that are probably confusing me, but I thought `||example.org^$dnstype=~CNAME` means that a domain is **not** blocked by CNAME or IP, and is only blocked when the domain itself is being requested. What am I missing out on here?

**Comment by ameshkov** [comment]
Well, that's not how `dnstype` works. You see, the type of the request is actually `A`, not `CNAME`, and that's why this modifier simply does not work as you expect it to even though it evaluates the `CNAME` record.

On the other hand, your problem seems absolutely valid to me. When evaluating response records, we should pass the record type, not the question type, to the filtering engine.

@ainar-g this should probably be categorized as enhancement, not a bug report. We'll also need to update the documentation as well.

**Reply by ainar-g** [reply]
@ameshkov, I'm not entirely sure I follow.  What should be the correct result here?

**Reply by ameshkov** [reply]
A new issue in the repo that will improve the `dnstype` semantics and cover the case when we apply rules to response records.

**Reply by ainar-g** [reply]
Ah, I see now, thanks.  I've filed #4238.

**Comment by EugeneOne1** [comment]
@DandelionSprout, hello again.  We've just pushed the edge build that should consider record types of responses instead of a question while filtering the response.  Could you please check, if it works properly?

**Reply by DandelionSprout** [reply]
Sorry for the late reply. My current AdGuard Home setup doesn't seem entirely stable, so I wasn't able to switch to the nightlies out of fear of breaking something. I'll apply the 0.108.0-b.3 beta pretty soon now, and will get the results from that in a couple days or so.
